csrc/optimization/remove_empty.cpp

```diff
@@ -0,0 +1,469 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <optimization/remove_empty.h>
+
+#include <ir/utils.h>
+#include <ops/alias.h>
+#include <ops/arith.h>
+
+#include <algorithm>
+#include <limits>
+#include <unordered_set>
+#include <vector>
+
+namespace nvfuser::optimization {
+
+namespace {
+
+//! This is a generic traversal class that is used to modify a Fusion graph by
+//! replacing TensorViews so that their definitions can be altered to simplify
+//! computation or remove dead code. This differs from OptOutMutator, which is
+//! built for mutating TensorViews in a graph, and does not easily handle
+//! modifying TensorView definitions and Expr Fusion inputs during traversal.
+//!
+//! We use unordered_set called live_statements_, which is initialized as the
+//! Exprs in traversal_exprs_ as well as their inputs and their outputs with
+//! live uses. Marking a Statement as dead removes it from live_statements_,
+//! and replacing a Val inserts the Val and its definition, recursively. Since
+//! we traverse backwards, and we handle all active Expr outputs, this ensures
+//! that it is safe to removing an Expr will not result in erasing definitions
+//! of active Expr outputs.
+//!
+//! Derived classes should override handle() and make use of replaceTV(),
+//! markDeadAndMaybeRemove(), and allUsesDead(). Note that if replacements are
+//! made using replaceTV(old_tv, new_tv), then neither new_tv or any new
+//! Statements produced in creating it will be traversed by this class.
+class DeadCodeRemover : BackwardVisitor {
```

Comment:
Question on the API. When to use `markDead` and when `markDeadAndMaybeRemove`? Any simple example case would be helpful.

Reply:
This was unclear. See latest commit. Child classes should now just use `removeVal` and `replaceVal`; marking dead and removal will happen automatically.